prov/sm2/src/sm2_ep.c

```diff
@@ -274,86 +246,59 @@ static void sm2_init_queue(struct sm2_queue *queue, dlist_func_t *match_func)
-	pend->iov_count = iov_count;
-	pend->peer_id = id;
-	pend->op_flags = op_flags;
-	pend->bytes_done = 0;
-	resp->status = FI_EBUSY;
-
-	pend->iface = iface;
-	pend->device = device;
-
-	resp->msg_id = (uint64_t)(uintptr_t)pend;
-}
-
-void sm2_generic_format(struct sm2_cmd *cmd, int64_t peer_id, uint32_t op, uint64_t tag,
-			uint64_t data, uint64_t op_flags)
-{
-	cmd->msg.hdr.op = op;
-	cmd->msg.hdr.op_flags = op == ofi_op_read_req ? SM2_RMA_REQ : 0;
-	cmd->msg.hdr.tag = tag;
-	cmd->msg.hdr.id = peer_id;
-	cmd->msg.hdr.data = data;
+	fqe->protocol_hdr.op = op;
+	fqe->protocol_hdr.op_flags = 0;
+	fqe->protocol_hdr.tag = tag;
+	fqe->protocol_hdr.id = self_id;
+	fqe->protocol_hdr.data = data;
 
 	if (op_flags & FI_REMOTE_CQ_DATA)
-		cmd->msg.hdr.op_flags |= SM2_REMOTE_CQ_DATA;
+		fqe->protocol_hdr.op_flags |= SM2_REMOTE_CQ_DATA;
 	if (op_flags & FI_COMPLETION)
-		cmd->msg.hdr.op_flags |= SM2_TX_COMPLETION;
+		fqe->protocol_hdr.op_flags |= SM2_TX_COMPLETION;
 }
 
-static void sm2_format_inject(struct sm2_cmd *cmd, enum fi_hmem_iface iface,
+static void sm2_format_inject(struct sm2_free_queue_entry *fqe, enum fi_hmem_iface iface,
 			      uint64_t device, const struct iovec *iov, size_t count,
-			      struct sm2_region *smr, struct sm2_inject_buf *tx_buf)
+			      struct sm2_region *smr)
 {
-	cmd->msg.hdr.op_src = sm2_src_inject;
-	cmd->msg.hdr.src_data = sm2_get_offset(smr, tx_buf);
-	cmd->msg.hdr.size = ofi_copy_from_hmem_iov(tx_buf->data, SM2_INJECT_SIZE, iface,
-						   device, iov, count, 0);
+	fqe->protocol_hdr.op_src = sm2_src_inject;
+	fqe->protocol_hdr.size = ofi_copy_from_hmem_iov(fqe->data, SM2_INJECT_SIZE, iface,
+							device, iov, count, 0);
 }
 
-int sm2_select_proto(bool use_ipc, bool cma_avail, enum fi_hmem_iface iface, uint32_t op,
-		     uint64_t total_len, uint64_t op_flags)
-{
-	return sm2_src_inject;
-}
+int sm2_select_proto() { return sm2_src_inject; }
 
-static ssize_t sm2_do_inject(struct sm2_ep *ep, struct sm2_region *peer_smr, int64_t id,
+static ssize_t sm2_do_inject(struct sm2_ep *ep, struct sm2_region *peer_smr,
 			     int64_t peer_id, uint32_t op, uint64_t tag, uint64_t data,
 			     uint64_t op_flags, enum fi_hmem_iface iface, uint64_t device,
-			     const struct iovec *iov, size_t iov_count, size_t total_len,
-			     void *context)
+			     const struct iovec *iov, size_t iov_count, size_t total_len)
 {
-	struct sm2_cmd *cmd;
-	struct sm2_inject_buf *tx_buf;
+	struct sm2_free_queue_entry *fqe;
+	struct sm2_region *self_region;
 
-	cmd = ofi_cirque_next(sm2_cmd_queue(peer_smr));
-	tx_buf = smr_freestack_pop(sm2_inject_pool(peer_smr));
+	assert(total_len <= 4096);
 
-	sm2_generic_format(cmd, peer_id, op, tag, data, op_flags);
-	sm2_format_inject(cmd, iface, device, iov, iov_count, peer_smr, tx_buf);
+	self_region = sm2_smr_region(ep, ep->self_fiaddr);
 
-	ofi_cirque_commit(sm2_cmd_queue(peer_smr));
-	peer_smr->cmd_cnt--;
+	if (smr_freestack_isempty(sm2_free_stack(self_region))) {
+		sm2_progress_recv(ep);
```

Comment:
Is there a significant performance improvement because of this change? If so, can you show it, please?

Reply:
The sender sends its own FQE's to the receiver.  The receiver processes the message and sends them back to the senders recv queue (dual purpose: receiving messages, and returning spent FQE's).  If the Sender only calls Send, and never calls progress, we will run out of FQE's to send.  This change ensures that if we are out of FQE's, we progress our receive to try and get more.  We then check again to see if we were able to get more, and if we weren't, we return -FI_EAGAIN, b/c the sender needs to wait until the other receivers progress and return its FQE's in order to send more (flow control).

Reply:
TLDR: This is a correctness change, not a performance change.